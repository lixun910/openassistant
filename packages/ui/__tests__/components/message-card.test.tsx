/**
 * @jest-environment jsdom
 */
import '@testing-library/jest-dom';

import React from 'react';
import { render, screen, fireEvent, act } from '@testing-library/react';
import MessageCard, {
  MessageCardProps,
} from '../../src/components/message-card';

describe('MessageCard Component', () => {
  const defaultProps: MessageCardProps = {
    index: 0,
    message: 'Test message',
    status: 'success',
    showFeedback: true,
  };

  it('renders the message correctly', () => {
    const { container } = render(<MessageCard {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it('displays the avatar when provided as a string', () => {
    const avatar = 'https://example.com/avatar.png';
    render(<MessageCard {...defaultProps} avatar={avatar} />);
    expect(screen.getByAltText('avatar')).toHaveAttribute('src', avatar);
  });

  it('displays the avatar when provided as a ReactNode', () => {
    const avatar = <div data-testid="custom-avatar">Custom Avatar</div>;
    render(<MessageCard {...defaultProps} avatar={avatar} />);
    expect(screen.getByTestId('custom-avatar')).toBeInTheDocument();
  });

  it('shows the failed message when status is failed', () => {
    render(<MessageCard {...defaultProps} status="failed" />);
    expect(
      screen.getByText(/Sorry, something went wrong/i)
    ).toBeInTheDocument();
    expect(screen.getByRole('link', { name: /Github/i })).toBeInTheDocument();
  });

  it('shows loading spinner when status is pending', () => {
    render(<MessageCard {...defaultProps} status="pending" />);
    expect(screen.getByTestId('spinner-icon')).toBeInTheDocument();
  });

  it('calls onMessageCopy when copy button is clicked', async () => {
    const onMessageCopy = jest.fn();
    await act(async () => {
      render(<MessageCard {...defaultProps} onMessageCopy={onMessageCopy} />);
    });
    const copyButton = screen.getByTestId('copytext-button');
    await act(async () => {
      fireEvent.click(copyButton);
    });
    expect(onMessageCopy).toHaveBeenCalledWith('Test message');
  });

  it('calls onFeedback when feedback button is clicked', async () => {
    const onFeedback = jest.fn();
    render(<MessageCard {...defaultProps} onFeedback={onFeedback} />);
    const feedbackButton = screen.getByTestId('feedback-button');
    await act(async () => {
      fireEvent.click(feedbackButton);
    });
    expect(onFeedback).toHaveBeenCalledWith(0);
  });

  it('renders attempt navigation when attempts > 1', async () => {
    const onAttemptChange = jest.fn();
    render(
      <MessageCard
        {...defaultProps}
        attempts={3}
        currentAttempt={2}
        onAttemptChange={onAttemptChange}
      />
    );
    const prevButton = screen.getByTestId('previous-button');
    const nextButton = screen.getByTestId('next-button');
    await act(async () => {
      fireEvent.click(prevButton);
    });
    expect(onAttemptChange).toHaveBeenCalledWith(1);
    await act(async () => {
      fireEvent.click(nextButton);
    });
    expect(onAttemptChange).toHaveBeenCalledWith(3);
  });

  it('renders attempt feedback when showFeedback is true and attempts > 1', async () => {
    const onAttemptFeedback = jest.fn();
    render(
      <MessageCard
        {...defaultProps}
        attempts={2}
        onAttemptFeedback={onAttemptFeedback}
      />
    );
    expect(
      screen.getByText(/Was this response better or worse?/i)
    ).toBeInTheDocument();
    const betterButton = screen.getByTestId('better-button');
    await act(async () => {
      fireEvent.click(betterButton);
    });
    expect(onAttemptFeedback).toHaveBeenCalledWith('like');
  });

  it('renders custom message when provided', () => {
    const customMessage = (
      <div data-testid="custom-message">Custom Message</div>
    );
    render(<MessageCard {...defaultProps} customMessage={customMessage} />);
    expect(screen.getByTestId('custom-message')).toBeInTheDocument();
  });

  it('renders screenshot when customMessage is a data URL', () => {
    const dataUrl =
      'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAACklEQVR4nGMAAQAABQABDQottAAAAABJRU5ErkJggg==';
    render(<MessageCard {...defaultProps} customMessage={dataUrl} />);
    const img = screen.getByRole('img', { name: /screenshot/i });
    expect(img).toHaveAttribute('src', dataUrl);
  });
});
