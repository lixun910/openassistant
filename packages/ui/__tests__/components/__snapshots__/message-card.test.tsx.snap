// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageCard Component renders the message correctly 1`] = `
<div>
  <div
    class="flex gap-3"
  >
    <div
      class="relative flex-none"
    >
      <div
        class="relative inline-flex shrink-0"
      >
        <span
          class="flex relative justify-center items-center box-border overflow-hidden align-middle z-0 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 w-10 h-10 text-tiny bg-default text-default-foreground rounded-full"
          tabindex="-1"
        >
          <span
            aria-label="avatar"
            class="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 flex items-center justify-center text-inherit w-full h-full"
            role="img"
          >
            <svg
              aria-hidden="true"
              fill="none"
              height="80%"
              role="presentation"
              viewBox="0 0 24 24"
              width="80%"
            >
              <path
                d="M12 2C9.38 2 7.25 4.13 7.25 6.75C7.25 9.32 9.26 11.4 11.88 11.49C11.96 11.48 12.04 11.48 12.1 11.49C12.12 11.49 12.13 11.49 12.15 11.49C12.16 11.49 12.16 11.49 12.17 11.49C14.73 11.4 16.74 9.32 16.75 6.75C16.75 4.13 14.62 2 12 2Z"
                fill="currentColor"
              />
              <path
                d="M17.0809 14.1489C14.2909 12.2889 9.74094 12.2889 6.93094 14.1489C5.66094 14.9989 4.96094 16.1489 4.96094 17.3789C4.96094 18.6089 5.66094 19.7489 6.92094 20.5889C8.32094 21.5289 10.1609 21.9989 12.0009 21.9989C13.8409 21.9989 15.6809 21.5289 17.0809 20.5889C18.3409 19.7389 19.0409 18.5989 19.0409 17.3589C19.0309 16.1289 18.3409 14.9889 17.0809 14.1489Z"
                fill="currentColor"
              />
            </svg>
          </span>
        </span>
        <span
          class="flex z-10 flex-wrap absolute box-border rounded-full whitespace-nowrap place-content-center origin-center items-center select-none font-regular scale-100 opacity-100 subpixel-antialiased data-[invisible=true]:scale-0 data-[invisible=true]:opacity-0 text-small px-0 border-2 border-background bg-danger text-danger-foreground w-5 h-5 min-w-5 min-h-5 bottom-[10%] right-[10%] translate-x-1/2 translate-y-1/2"
          data-invisible="true"
        >
          <span />
        </span>
      </div>
    </div>
    <div
      class="flex w-full flex-col gap-4 overflow-x-auto"
    >
      <div
        class="group relative w-full rounded-medium px-4 py-3 
              
              undefined"
      >
        <div
          class="min-h-8 whitespace-pre-line max-w-full"
        >
          Test message
        </div>
        <div
          class="opacity-0 group-hover:opacity-100 absolute right-2 top-2 flex rounded-full shadow-small bg-content2"
        >
          <div
            class=""
          >
            <button
              class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-tiny gap-2 rounded-full px-0 !gap-0 transition-transform-colors-opacity motion-reduce:transition-none bg-transparent text-default-foreground data-[hover=true]:bg-default/40 min-w-8 w-8 h-8"
              data-testid="drag-button"
              draggable="false"
              type="button"
            >
              <span />
            </button>
            <button
              class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-tiny gap-2 rounded-full px-0 !gap-0 transition-transform-colors-opacity motion-reduce:transition-none bg-transparent text-default-foreground data-[hover=true]:bg-default/40 min-w-8 w-8 h-8"
              data-testid="copytext-button"
              type="button"
            >
              <span />
            </button>
            <button
              class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-tiny gap-2 rounded-full px-0 !gap-0 transition-transform-colors-opacity motion-reduce:transition-none bg-transparent text-default-foreground data-[hover=true]:bg-default/40 min-w-8 w-8 h-8"
              data-testid="feedback-button"
              type="button"
            >
              <span />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
